### PR TITLE
Implementation of the `expect` attribute (RFC 2383)

### PR DESCRIPTION
**Pull request by xFrednet**
This is an implementation of the `expect` attribute as described in [RFC-2383](https://rust-lang.github.io/rfcs/2383-lint-reasons.html). The attribute allows the suppression of lint message by expecting them. Unfulfilled lint expectations (meaning no expected lint was caught) will emit the `unfulfilled_lint_expectations` lint at the `expect` attribute.

#### input
```rs
// required feature flag
#![feature(lint_reasons)]

#[expect(unused_mut)] // Will warn about an unfulfilled expectation
#[expect(unused_variables)] // Will be fulfilled by x
fn main() {
    let x = 0;
}
```

#### output

```txt
warning: this lint expectation is unfulfilled
  --> $DIR/trigger_lint.rs:3:1
   |
LL | #[expect(unused_mut)] // Will warn about an unfulfilled expectation
   |          ^^^^^^^^^^
   |
   = note: `#[warn(unfulfilled_lint_expectations)]` on by default
```

### Implementation

This implementation introduces `Expect` as a new lint level for diagnostics, which have been expected. All lint expectations marked via the `expect` attribute are collected in the [`LintLevelsBuilder`] and assigned an ID that is stored in the new lint level. The `LintLevelsBuilder` stores all found expectations and the data needed to emit the `unfulfilled_lint_expectations` in the [`LintLevelsMap`] which is the result of the [`lint_levels()`] query. 

The [`rustc_errors::HandlerInner`] is the central error handler in rustc and handles the emission of all diagnostics. Lint message with the level `Expect` are suppressed during this emission, while the expectation ID is stored in a set which marks them as fulfilled. The last step is then so simply check if all expectations collected by the [`LintLevelsBuilder`] in the [`LintLevelsMap`] have been marked as fulfilled in the [`rustc_errors::HandlerInner`]. Otherwise, a new lint message will be emitted.

The implementation of the `LintExpectationId` required some special handling to make it stable between sessions. Lints can be emitted during [`EarlyLintPass`]es. At this stage, it's not possible to create a stable identifier. The level instead stores an unstable identifier, which is later converted to a stable `LintExpectationId`.

### Followup TO-DOs

Moved to: https://github.com/rust-lang/rust/issues/85549

### Open questions
I also have a few open questions where I would like to get feedback on:
1. The RFC discussion included a suggestion to change the `expect` attribute to something else. (Initiated by @Ixrec [here](https://github.com/rust-lang/rfcs/pull/2383#issuecomment-378424091), suggestion from @scottmcm to use `#[should_lint(...)]` [here](https://github.com/rust-lang/rfcs/pull/2383#issuecomment-378648877)). No real conclusion was drawn on that point from my understanding. Is this still open for discussion, or was this discarded with the merge of the RFC?
2. How should the expect attribute deal with the new `force-warn` lint level?

---

This approach was inspired by a discussion with @LeSeulArtichaut.

RFC tracking issue: #54503

Mentoring/Implementation issue: #85549

[`LintLevelsBuilder`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/levels/struct.LintLevelsBuilder.html
[`LintLevelsMap`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/lint/struct.LintLevelMap.html
[`lint_levels()`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.lint_levels
[`rustc_errors::HandlerInner`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.HandlerInner.html
[`EarlyLintPass`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/trait.EarlyLintPass.html

### CODE REVIEW

**Comment by LeSeulArtichaut** [comment]
I don't think I can help much with this. You might want to add a test for macros however.

**Comment by xFrednet** [comment]
That's a very good pointer, thank you for taking a look :upside_down_face:

**Comment by xFrednet** [comment]
Hello, I believe that this PR is ready for review. There is definitely one To-Do that has to be addressed before the merge when it comes to incremental compilation and the stability of the `LintExpectationId`. However, I would first like to get some feedback if the structure implementation is good before diving further into this.

@flip1995 You offered in the last PR to aid with the review. I would appreciate it if you could take a look at this one and maybe give some feedback :upside_down_face:.

r? @wesleywiser Would you be willing to take over the main review of this? :upside_down_face:

File: src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs
Comment (flip1995) [inline src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs:29]: This should trigger the unfulfilled lint for `unused_mut`. The reason it doesn't is that `unused` is globally allowed in compiletest. If you add a `#![warn(unused_mut`)]` at the top of the file, this should fail.

Maybe add a general test case where `expect` overrides `allow`/`warn` similar to the above tests where it's the other way around.

File: src/test/ui/lint/rfc-2383-lint-reason/expect_lint_from_macro.stderr
Comment (xFrednet) [reply src/test/ui/lint/rfc-2383-lint-reason/expect_lint_from_macro.stderr:28]: Yes it should, the second warning is weirdly enough triggered by the `check_expect_on_macro()` function in the test. I'll investigate this :upside_down_face: 

```rs
pub fn check_expect_on_macro() {
    // This should be fulfilled by the macro
    #[expect(unused_variables)]
    trigger_unused_variables_macro!();
}
```

File: src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs
Comment (xFrednet) [reply src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs:29]: > This should trigger the unfulfilled lint for `unused_mut`. The reason it doesn't is that `unused` is globally allowed in compiletest. If you add a `#![warn(unused_mut`)]` at the top of the file, this should fail.

The RFC document doesn't seem to define what happens if multiple lints have been listed, so it's probably open for discussion. This implementation sees the `expect` attribute as one expectation that can be fulfilled by any of the listed lints (This is also what the test should illustrate). I found this reasonable for an example like this:
```rs
#[expect(unused_mut, unused_value, reason = "This is work in progress, these are expected to happen")]
```
Here the reason stands for all of them and I don't care which of them is triggered. If I would like to be more precise, I can simply add more expect attributes. This gives the user more freedom IMO.

File: compiler/rustc_lint/src/expect.rs
Comment (xFrednet) [reply compiler/rustc_lint/src/expect.rs:10]: The [`lint_reasons` feature documentation](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_feature/active/struct.Features.html#structfield.lint_reasons) indicates that it should be used for `expect` attribute as well. However, we can most likely add another feature for this without any problems, if we wanted to. I don't really have an opinion on this.

File: src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs
Comment (flip1995) [reply src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs:29]: Hm, that would mean, if I expect multiple lints, I would have to add multiple `expect` attributes. This seems really counter-intuitive to me. `allow(A, B)` allows the lints `A` and `B`, so `expect(A, B)` should also expect `A` and `B` and not `A` _or_ `B`.

File: compiler/rustc_lint_defs/src/lib.rs
Comment (xFrednet) [reply compiler/rustc_lint_defs/src/lib.rs:56]: The explanation is a bit more complicated, but here is the gist of it:

From my understanding, this is required for two reasons:
1. The incremental compilation implementation stores diagnostics between compilations on disc and replays them if the producing query has been marked as green (meaning that it won't be executed). This means that a diagnostic can be emitted with a `LintExpectationId` that originated from the last session. (I'm not totally sure on the rules, it might be that this is not actually a problem. The question would be if any of the linting query can be marked as green when any expectations have been changed for that crate/module)
2. The expectations are collected during the [`lint_levels`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.lint_levels) query and stored in the [`LintLevelMap`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/lint/struct.LintLevelMap.html) struct. The query is marked as `eval_always` but my understanding is that the result is then hashed and compared with the stored hash. The `LintExpectationId` is part of the query result and therefore also used to check if the query result has changed. Having it unstable can result in a different hash for the same query input.

File: src/test/ui/lint/rfc-2383-lint-reason/expect_nested_lint_levels.rs
Comment (xFrednet) [reply src/test/ui/lint/rfc-2383-lint-reason/expect_nested_lint_levels.rs:20]: Yes, this can be confusing at the start. The RFC stated:
> This RFC also adds the `expect(lint_name, reason = STRING)` lint attribute. The
> `expect` attribute uses the same lint-suppression mechanism that `allow` does,
> but will raise a new lint, `expectation_missing` (name open to change), when the
> lint it expects does not arrive.

The *"same lint-suppression mechanism that `allow` does"* part makes it sounds like this was expected to be like another lint level. The RFC discussion also hinted to the fact. I'll update the reason parameter to explain this better :upside_down_face:

File: src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs
Comment (xFrednet) [reply src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs:29]: Yeah, I can see the reasoning behind this. I'll give it some more thought, but probably change the implementation. The change can be done by giving each listed lint it's on `LintExpectationId`. That again touch on the question of how that ID should be generated. I'll ask the incremental compilation Zulip stream for some thoughts on the ID before applying this change :upside_down_face:.

Thank you for the feedback!

File: src/test/ui/lint/rfc-2383-lint-reason/expect_lint_from_macro.stderr
Comment (xFrednet) [reply src/test/ui/lint/rfc-2383-lint-reason/expect_lint_from_macro.stderr:28]: It turns out that this is a bug, but it's not exclusive to the `expect` attribute. The same thing happens with the `allow` attribute. See [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3af93c39e149559ebd6029c2005d58f2) and issue #87391 mentioning this bug as well. I would remove the test and add a `FIXME` comment to the file. Would that work for now? :upside_down_face:

File: src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs
Comment (xFrednet) [reply src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs:29]: > If you add a `#![warn(unused_mut)]` at the top of the file, this should fail.

I've now also added the corresponding `warn` attributes for the lints in question to all test files :upside_down_face:

File: src/test/ui/lint/rfc-2383-lint-reason/expect_lint_from_macro.stderr
Comment (flip1995) [reply src/test/ui/lint/rfc-2383-lint-reason/expect_lint_from_macro.stderr:28]: I would leave the test in it and add a fixme comment on the test case.

**Comment by flip1995** [comment]
Something that came to my mind this morning under the shower: What if you expect an EarlyLintPass and a LateLintPass lint at the same item. So:

```rs
#[expect(some_early_lints)]
#[expect(some_late_lints)]
fn foo() {
    // code triggering all specified lints
}
```

I think  this isn't a problem after #87337 gets merged, but I would add a test for it anyway.

**Comment by xFrednet** [comment]
> What if you expect an EarlyLintPass and a LateLintPass lint at the same item. 

That's a good thought. The expectations are currently checked at the end of the last lint pass, meaning that it should still work as expected. Adding a test for it is still a very good idea. I'll look up some lints for a test like that.

---

What are your thoughts on an example where an item has multiple expect attributes with the same lint? For instance
```rs
#[expect(unused_variables)] // 1
#[expect(unused_variables)] // 2
fn main() {
    let x = 0;
}
```

A test with normal lint levels shows that they are processed in order top-down ([example playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ff6405efdf90585c0f31b46521cb1609)). This means that the current implementation would mark the second expectation as fulfilled, and lint about the first one. Does this seem reasonable? :upside_down_face:

**Comment by flip1995** [comment]
> This means that the current implementation would mark the second expectation as fulfilled, and lint about the first one. Does this seem reasonable? :upside_down_face:

Intuitively I would expect no warning in this case, if adding 2 identical `allow` attributes don't produce a warning either. The current behavior is fine for now as long as it is unstable, but before stabilizing it this is an open question that should be addressed.

**Comment by xFrednet** [comment]
This PR is still blocked on the question of how to stabilize the expect-ID. See [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/241847-t-compiler.2Fwg-incr-comp/topic/implementing.20the.20expect.20RFC.20rust.2387835/near/248963523), the question sadly didn't receive any answers.

I'm hoping that I'll be able to stabilize the ID after the work that is currently being done on the lowering queries (I only heard of them and haven't yet looked at them in detail)

@rustbot label +S-waiting-on-author +S-blocked -S-waiting-on-review

**Comment by crlf0710** [comment]
`S-blocked` is not for this case. I've added `I-needs-decision` label.

File: src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs
Comment (xFrednet) [reply src/test/ui/lint/rfc-2383-lint-reason/fulfilled_expect.rs]: The latest update now tracks each lint individually. :upside_down_face:

**Comment by xFrednet** [comment]
Hey, I finally found a solution to have a *stable* `LintExpectationId` by having it unstable at first and then converting it to a stable version once all notes have `HirId`s. I believe that this is a working implementation of the `expect` attributes. The open to-do listed in the PR description are still open but can be done in followup PRs. This is for review now. :upside_down_face:

This PR increases the size of `Level` enums which are part of every diagnostic message. It might be good to run a performance test to ensure that this doesn't have an impact on performance. The `LintExpectationId` can be reduced by an additional 4 bytes by changing the index types to `u16` values instead of `u32`. The last commit has already changes the type to an `u32` instead of the original `usize`.

@rustbot label -I-needs-decision
@rustbot ready

**Comment by Aaron1011** [comment]
Can you add a test for combining `#[expect]` and `#[forbid]`? The RFC doesn't specifically mention that interaction, but it does state that `expect` "functions identically to `allow`"`. I think the correct behavior is to emit an error as if the `#[expect]` attribute was not present.

**Comment by xFrednet** [comment]
> Can you add a test for combining `#[expect]` and `#[forbid]`? The RFC doesn't specifically mention that interaction, but it does state that `expect` "functions identically to `allow`"`. I think the correct behavior is to emit an error as if the `#[expect]` attribute was not present.

Nice suggestion, the latest commit added a test for this :upside_down_face:

// Let's rerun these, I though that all bors test would be covered by the normal pipeline
@bors try @rust-timer queue

**Comment by xFrednet** [comment]
`inflate` is the only test with a large performance hit, the other regressions (< 1% difference) were kind of expected. I've now reduced the size of `LintExpectationId` by 4 bytes and moved some calculations from the `LintLevelsBuilder` into the new lint emission. Let's see if this made a difference:

@bors try @rust-timer queue

**Comment by xFrednet** [comment]
Alright, I think this performance test looks reasonable. It was expected that this addition will have some slight regressions, the amount < 0.4% seems okay to me. This is now finally ready for review form myside.

@rustbot label -S-waiting-on-author

File: src/test/ui/lint/rfc-2383-lint-reason/expect_with_forbid.rs
Comment (Aaron1011) [inline src/test/ui/lint/rfc-2383-lint-reason/expect_with_forbid.rs:6]: Could you add another test that has the attributes in the order `#[expect] #[forbid]` ?

File: compiler/rustc_lint/src/expect.rs
Comment (Aaron1011) [inline compiler/rustc_lint/src/expect.rs:19]: Can you add a panic if this check fails?

File: compiler/rustc_errors/src/lib.rs
Comment (Aaron1011) [inline compiler/rustc_errors/src/lib.rs:970]: If I understand correctly, this should always succeed, since you only push "expect"-level diagnostics to `unstable_expect_diagnostics`. Could you add an `else` branch with an panic to ensure that this branch is always actually taken?

File: compiler/rustc_errors/src/lib.rs
Comment (Aaron1011) [inline compiler/rustc_errors/src/lib.rs:976]: Same thing here, I think - `unstable_id` should always be successfully mapped here, so you could use `.expect` to enforce this.

Comment (xFrednet) [reply compiler/rustc_errors/src/lib.rs]: You're correct with all assumptions, I've added the corresponding tests. Thank you!

File: src/test/ui/lint/rfc-2383-lint-reason/expect_with_forbid.rs
Comment (xFrednet) [reply src/test/ui/lint/rfc-2383-lint-reason/expect_with_forbid.rs:6]: The tests have been added. Then it simply works, like it would for `#[allow]`. The output also contains an `unfulfilled_lint_expectation` message which is expected as the forbidden diagnostic doesn't fulfill the expectation.

**Comment by xFrednet** [comment]
(Rebased and squashed some commits)

Hey @wesleywiser, this PR is ready for review. I've requested you as a reviewer since you were also assigned to the original PR. Are you willing to take over this review, or should I search for someone else? :upside_down_face: 

@rustbot ready

**Comment by wesleywiser** [comment]
Hey @xFrednet! I've started reviewing but haven't finished yet. Thanks for your patience!

**Comment by xFrednet** [comment]
Thank you very much for your response, I wasn't certain if you were still following this PR. I highly appreciate it :upside_down_face:

File: compiler/rustc_errors/src/lib.rs
Comment (wesleywiser) [inline compiler/rustc_errors/src/lib.rs:460]: Perhaps we should assert this is empty in `steal_fulfilled_expectation_ids`?

File: compiler/rustc_lint/src/levels.rs
Comment (wesleywiser) [inline compiler/rustc_lint/src/levels.rs:250]: I think this would be clearer if we constructed the `LintExpecationId` before calling `Level::from_symbol` and changed that function back to take a plain `LintExceptationId` instead of a closure. Something like

```rust
let expectation_id = if let Some(hir_id) = source_hir_id {
    LintExpecationId::Stable { hir_id, attr_index }
} else { 
    LintExpecatationId::Unstable(attr.id)
};
let Some(level) = Level::from_symbol(attr.name_or_empty(), expectation_id) else {
    continue;
}
```

If we can use the stable id from the get-go, I don't see why we would need to create an unstable id and then register the stable id in the `expecatation_id_map` right after the fact but maybe I'm missing something 🙂

File: src/test/ui/lint/rfc-2383-lint-reason/catch_multiple_lint_triggers.rs
Comment (wesleywiser) [inline src/test/ui/lint/rfc-2383-lint-reason/catch_multiple_lint_triggers.rs:1]: Note for me to review the UI tests more thoroughly at the end of the commit set.

File: compiler/rustc_errors/src/lib.rs
Comment (wesleywiser) [inline compiler/rustc_errors/src/lib.rs]: Noting for myself that I stopped reviewing at this commit.

**Review by wesleywiser — COMMENTED** [review]
Still working on the full review, sorry for the slow progress!

**Comment by wesleywiser** [comment]
Small procedural note: if you want to fix anything I mentioned in the review comments, please feel free but I would ask that you push them as additional commits in the branch and not rebase the current branch until the review is finished. That will make it much easier for me to keep track of where in the review I am. 

Thanks!

File: compiler/rustc_lint/src/expect.rs
Comment (xFrednet) [reply compiler/rustc_lint/src/expect.rs]: Are you referring to the double spaces behind the `FIXME` or is this just a reminder that it is there? :thinking:

File: compiler/rustc_errors/src/lib.rs
Comment (xFrednet) [reply compiler/rustc_errors/src/lib.rs:460]: That's probably a good idea :+1:

File: compiler/rustc_lint/src/levels.rs
Comment (xFrednet) [reply compiler/rustc_lint/src/levels.rs:250]: The `expecatation_id_map` is needed to map from unstable to stable IDs. If we move it in front of the `match` statement, we would need an additional check to add the unstable to stable mapping if the level is `Expect`. We could also add each mapping to the list, but I would like to avoid it, as this code runs for every attribute. I'm guessing that this match statement is the most efficient, however that's just a guess.

If you have a suggestion where the ID mapping should be added or how it can be changed, I could run a benchmark :upside_down_face:

**Comment by xFrednet** [comment]
Thank you for the partial review. The comments are minor enough that I'll wait until you're done with the complete review :upside_down_face:

**Comment by xFrednet** [comment]
Hey @wesleywiser, have you had the time to review this a bit further? :upside_down_face: I'm waiting for you to complete the review until I rebase on master to resolve the conflict. So far, they are luckily easy to fix.

Also on a related note, I expect that the changes from #93368 will cause some conflicts, but nothing major.

**Comment by nikomatsakis** [comment]
Oh, I'm super excited to see an implementation of `#[expect]`! I love the dead-code lint, but it can be so frustrating when just getting code up-and-running for the first time.

**Comment by xFrednet** [comment]
I've cleaned up the PR a bit by hiding resolved and outdated comments. I hope that this simplifies the review. :)

**Comment by wesleywiser** [comment]
That's super helpful, thanks @xFrednet! I haven't been able to find much time over the last few weeks do larger, in depth reviews like this requires but I have scheduled Monday to try to complete this review. Sorry for how long this has taken!

**Comment by xFrednet** [comment]
Hey @Dylan-DPC, thank you for the hint. The current conflicts are all simple to fix. (A function signature has changed, and a few imports were added). `@wesleywiser` [asked](https://github.com/rust-lang/rust/pull/87835#issuecomment-995176632) not to rebase until he is finished with his review. I'm waiting on a comment from him regarding this before I do so. :upside_down_face:

**Review by wesleywiser — APPROVED** [review]
Overall, this looks really good to me! I'd like to go over a few parts again tomorrow and think a bit more about them, but I'm very happy with the state of the code as it is currently. Thanks for your patience getting this reviewed!

**Comment by Dylan-DPC** [comment]
> Is it alright with you if I address your new comments and then rebase to resolve the conflicts? upside_down_face

yeah that should be fine

**Comment by wesleywiser** [comment]
Oh, yes, please do! I should have mentioned that in my review comment 🙂

File: compiler/rustc_errors/src/lib.rs
Comment (xFrednet) [reply compiler/rustc_errors/src/lib.rs]: I changed the comment to make clear what I wanted to say :)

```rs
// The unstable to stable map only maps the unstable `AttrId` to a stable `HirId` with an attribute index.
// The lint index inside the attribute is manually transferred here.
```

**Comment by xFrednet** [comment]
Alright, I've rebased and everything is green, the conflicts were luckily quite simple to solve. You are welcome to reach out if anything else should be changed in this PR or if I can help with anything :upside_down_face:

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
This is really great work @xFrednet! I'm sorry it's taken so long for me to finish reviewing.

> Followup TO-DOs

I think we should land this PR and make those changes in follow ups. As for `#[expect(unfulfilled_lint_expectations )]`, the behavior [mentioned in the RFC PR](https://github.com/rust-lang/rfcs/pull/2383#issuecomment-397468421) seems extremely subtle and I'm having trouble imagining a use case for it. We may want to consider making `#[expect(unfulfiled_lint_expectations)]` an error for now. (The comments I linked to in the RFC PR also don't seem to have been incorporated into the RFC text so it's unclear to me if they are the expected behavior or not). 

> Open Questions

1. We should raise this point in the tracking issue to verify that `expect` is the correct name for the lint level.
2. I *think* based on reading the documentation for `force-warn`, `expect` should not be able to catch any lints generated at the `force-warn` level but we should also raise this question in the tracking thread. 

I don't think any of these need to block landing this PR so if that all sounds good to you, I'll r+!

**Comment by xFrednet** [comment]
Thank you very much @wesleywiser! I'm also pleased with how all of this turned out. The break in between was also good as I was writing my bachelor thesis in that time. :upside_down_face:

> I think we should land this PR and make those changes in follow ups.

Agreed! I think the behavior of `#[expect(unfulfilled_lint_expectations)]` is not really needed and was mostly a general question about how diagnostics should cascade. I'll think a bit more about it and then create a follow-up PR to handle that case. (Either with an error or another solution)

> I don't think any of these need to block landing this PR so if that all sounds good to you, I'll r+!

Sounds perfect to me! I'll ask these questions in the issue. Should I also ping someone like `@rust-lang/compiler` regarding the questions?

**Comment by flip1995** [comment]
As for `#[expect(unfulfilled_lint_expectations)]` I think adding this to the code, is extremely useless and confusing when reading the code.

I think the easier case is, if `#[expect(unfulfilled_lint_expectations)]` is in the code, but the `unfulfilled_lint_expectations` doesn't trigger in its scope. In that case the lint should trigger on `#[expect(unfulfilled_lint_expectations)]` and suggest to remove it.

The other case, where `#[expect(unfulfilled_lint_expectations)]` is added and the `unfulfilled_lint_expectations` does trigger in its scope, is a bit harder.

In that case, you are pretty much saying, that you expect that an `expect` attribute can be removed, because the lint won't trigger in the scope where the `expect` attribute is attached to. Instead of doing this, you should rather remove the offending `expect` attribute or turn it into an `allow` attribute. 

So maybe the `unfulfilled_lint_expectations` (or a separate) lint should trigger if it finds a `#[expect(unfulfilled_lint_expectations)]` and suggest to remove the offending `expect` attribute or change it to `allow`.

I don't think we should hard-error in the final version of the `expect` attribute though.

I'm not concerned about paying attention of the placement of the attribute and how it behaves differently in those cases, as raised in the RFC comment. This should Just Work (tm) with how the `push`/`pop` functions for lint attributes work.

---

> I _think_ based on reading the documentation for `force-warn`, `expect` should not be able to catch any lints generated at the `force-warn` level but we should also raise this question in the tracking thread.

IMO, a `force-warn` lint should trigger regardless whether there's an `expect` attribute. 

> If a lint is set to 'force-warn', it is guaranteed to warn: no more, no less. This is true even if the overall lint level is capped via cap-lints.

I think the documentation is pretty clear about that.

**Comment by xFrednet** [comment]
> This should Just Work (tm) with how the push/pop functions for lint attributes work.

The current implementation will not directly work as the diagnostic is created after all fulfilled lint expectations have been stolen. However, it should be simple to add that behavior

>> If a lint is set to 'force-warn', it is guaranteed to warn: no more, no less. This is true even if the overall lint level is capped via cap-lints.
>
> I think the documentation is pretty clear about that.

True, I also came to the same conclusion just now ^^. The question next is, should the expectation be marked as fulfilled, even if the lint level has been set to warn? I believe it should, as the lint was still triggered inside the marked scope. `force-warn` was introduced to work with `rust fix` which also supports this behavior. (See: https://github.com/rust-lang/rust/issues/85512)

**Comment by xFrednet** [comment]
The usage of `FxHashMap` made a few problems as the expectation check order was not constant. I've rebased and added a new commit to use a `Vec` instead. The other files stayed the same :upside_down_face: 

@bors try

**Pull request merged**
